src/tests/test_courses.py

```diff
@@ -611,15 +611,20 @@ def test_course_completion_endpoint(self, user_client, user):
         Тест эндпоинта завершения курса.
 
         Проверяем, что при завершении курса, пользователь получает статус COMPLETED.
+        Проверяем что курс можно завершить только с завершенными главами, в противном случае получаем 403.
         """
         course = CourseWith2Chapters()
-        _ = SubscriptionFactory(user=user, course=course)
+        user_completed_chapters = user.user_chapter_status.all().filter(progress=2, chapter__course=course.pk)
+        total_completed_chapters = course.chapters.all()
         url = reverse("courses-complete", kwargs={"pk": course.id})
         url_course = reverse("courses-detail", kwargs={"pk": course.id})
         response = user_client.post(url)
         response_course = user_client.get(url_course)
-        assert response.status_code == status.HTTP_200_OK
-        assert response_course.json()["user_status"] == Subscription.Status.COMPLETED
+        if len(user_completed_chapters) == len(total_completed_chapters):
+            assert response.status_code == status.HTTP_200_OK
+            assert response_course.json()["user_status"] == Subscription.Status.COMPLETED
+        else:
+            assert response.status_code == status.HTTP_403_FORBIDDEN
```

Comment:
Нам не важна внутрення логика, нам надо проверить только 2 вещи:
1. Курс не завершается если не все уроки пройдены
2. Курс завершается если все уроки пройдены
```suggestion
        url = reverse("courses-complete", kwargs={"pk": course.id})
        url_course = reverse("courses-detail", kwargs={"pk": course.id})
        response = user_client.post(url)
        assert response.status_code == status.HTTP_403_FORBIDDEN

        lessons = Lesson.objects.filter(chapter_course=course)
        for lesson in lessons:
            lesson.finish(user)
        
        response = user_client.post(url)
        response_course = user_client.get(url_course)
        assert response.status_code == status.HTTP_200_OK
        assert response_course.json()["user_status"] == Subscription.Status.COMPLETED
```

Reply:
Не совсем понял о чем тут имеется ввиду. 
стр. №624-627 как раз проверяют если кол-во завершенных глав соответствует кол-ву глав в курсе то COMPLETED & 200. В противном случае 403 ?

Reply:
Это то что нужно было сделать ? 
`    def test_course_completion_endpoint(self, user_client, user):
        """
        Тест эндпоинта завершения курса.

        Проверяем, что при завершении курса, пользователь получает статус COMPLETED.
        Проверяем что курс можно завершить только с завершенными главами, в противном случае получаем 403.
        """
        course = CourseWith2Chapters()
        url = reverse("courses-complete", kwargs={"pk": course.id})
        url_course = reverse("courses-detail", kwargs={"pk": course.id})
        response = user_client.post(url)
        assert response.status_code == status.HTTP_404_NOT_FOUND
        subscription = SubscriptionFactory(user=user, course=course)
        response = user_client.post(url)
        assert response.status_code == status.HTTP_403_FORBIDDEN
        response_course = user_client.get(url_course)
        lessons = Lesson.objects.filter(chapter__course=course)
        for lesson in lessons:
            lesson.finish(subscription)
        response = user_client.post(url)
        response_course = user_client.get(url_course)
        assert response.status_code == status.HTTP_200_OK
        assert response_course.json()["user_status"] == Subscription.Status.COMPLETED`